docs/syntax-attribute-functions-deprecated.md

```diff
@@ -0,0 +1,60 @@
+# Deprecated Attribute for Functions
+
+## Summary
+
+This RFC proposes a `@deprecated` attribute to annotate Luau functions for deprecation. Using these functions will result in a deprecation warning from the linter. The Luau LSP can indicate the deprecated status of these functions in its completion list by showing them in a visually distinct style.
+
+## Motivation
+
+Roblox already has numerous deprecated Luau APIs. LSP strikes out the names of these APIs in the autocompletion popup in Roblox Studio, their use issues deprecation warnings, and their deprecation status is reflected in Roblox's Creator Hub documentation. Unfortunately, Luau lacks support to leverage this tooling for deprecating custom creator-defined functions. Even for core Luau deprecated functions, `getfenv` and `setfenv`, the Linter has hardcoded name comparisons to identify and report deprecation messages. Deprecating new APIs requires ad-hoc modifications to the tooling which can only be performed by the language designers.
+
+This RFC proposes the introduction of a `@deprecated` attribute to mark deprecated Luau APIs. This has multiple benefits:
+1. It provides an explicit syntactic marker to identify deprecated functions in Luau code.
+2. The attribute can be read by LSP and the corresponding API can be displayed in a visually distinct style in the autocompletion popup.
+3. Linter can issue warning messages when the deprecated function is used.
+
+## Design
+
+The `@deprecated` attribute can take upto two named string parameters. The parameters affect the warning message issued by the tooling. The `msg` parameter is the warning message issued by the tooling when it identifies a use of the deprecated function. This should typically include the reason for deprecation. The `alt` parameter is the name of the function that should be used in lieu of the deprecated function. The following table shows the messages issued by the different styles of this attribute when used on a function named `foo`.
+
+| Style                                  | Message                                                                |
+| -------------------------------------- | -----------------------------------------------------------------------|
+| `@deprecated`                          | `"Function 'foo' is deprecated."`                                      |
+| `@[deprecated {msg = msg}]`            | `"Function 'foo' is deprecated. <msg>"`                                |
+| `@[deprecated {alt = alt}]`            | `"Function 'foo' is deprecated. Use '<alt>' instead."`                 |
+| `@[deprecated {msg = msg, alt = alt}]` | `"Function 'foo' is deprecated. <msg> Use '<alt>' instead.`            |
+
+The `@deprecated` attribute can only be used on top-level functions. It does not apply recursively to the functions defined within the lexical scope of the attributed function.
```

Comment:
Do not see the need to restrict to top level.

Reply:
Conceptually, the analysis feels simpler (tractable?) when you restrict to top-level functions. Also, I would be surprised if inner functions are used as the public API of a module. Having said this, I would not mind generalizing it to functions at any nesting depth.

Reply:
Deprecation status is already a field on `Luau::FunctionType`, so it doesn't actually affect the implementation effort.